modules/reindex/src/main/java/org/elasticsearch/index/reindex/ReindexTaskStateUpdater.java

```diff
@@ -154,28 +155,42 @@ public void onFailure(Exception e) {
     public void finish(@Nullable BulkByScrollResponse reindexResponse, @Nullable ElasticsearchException exception) {
         assert checkpointThrottler != null;
         if (isDone.compareAndSet(false, true)) {
-            checkpointThrottler.close(() -> writeFinishedState(reindexResponse, exception));
+            checkpointThrottler.close(() -> writeFinishedState(reindexResponse, exception, TimeValue.ZERO));
         }
     }
 
-    private void writeFinishedState(@Nullable BulkByScrollResponse reindexResponse, @Nullable ElasticsearchException exception) {
+    private void writeFinishedState(@Nullable BulkByScrollResponse reindexResponse, @Nullable ElasticsearchException exception,
+                                    TimeValue delay) {
         ReindexTaskStateDoc state = lastState.getStateDoc().withFinishedState(reindexResponse, exception);
         long term = lastState.getPrimaryTerm();
         long seqNo = lastState.getSeqNo();
+        lastState = null;
+
         reindexIndexClient.updateReindexTaskDoc(persistentTaskId, state, term, seqNo, new ActionListener<>() {
             @Override
             public void onResponse(ReindexTaskState taskState) {
-                lastState = null;
                 finishedListener.onResponse(taskState.getStateDoc());
 
             }
 
             @Override
             public void onFailure(Exception e) {
-                lastState = null;
-                // TODO: Maybe retry finished write?
-                finishedListener.onFailure(e);
+                // TODO: Need to ensure that the allocation has changed
+                if (e instanceof VersionConflictEngineException == false) {
+                    TimeValue nextDelay = getNextDelay(delay);
+                    threadPool.schedule(() -> writeFinishedState(reindexResponse, exception, nextDelay), nextDelay, ThreadPool.Names.SAME);
+                }
             }
         });
     }
+
+    private TimeValue getNextDelay(TimeValue delay) {
+        TimeValue newDelay;
+        if (TimeValue.ZERO.equals(delay)) {
+            newDelay = TimeValue.timeValueMillis(500);
+        } else {
+            newDelay = TimeValue.timeValueSeconds(Math.max(delay.getMillis() * 2, TimeValue.timeValueMinutes(1).millis()));
```

Comment:
`timeValueSeconds` looks like it should be `timeValueMillis`?

I think we should raise the max to something like 30 minutes.

As a follow-up, we should consider observing the cluster state and retrigger whenever a shard of the index goes from RED to YELLOW (or something similar)?